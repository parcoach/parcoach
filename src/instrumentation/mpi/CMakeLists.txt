add_sources_to_format(SOURCES MPI_DynamicCheck.c)
add_library(MPI_DynamicCheck OBJECT MPI_DynamicCheck.c)
set_property(TARGET MPI_DynamicCheck PROPERTY POSITION_INDEPENDENT_CODE 1)
target_link_libraries(MPI_DynamicCheck PRIVATE MPI::MPI_C)

target_link_libraries(Instrumentation MPI_DynamicCheck)
