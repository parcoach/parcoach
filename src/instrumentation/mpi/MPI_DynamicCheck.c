#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>

#include "mpi.h"

int NbCc = 0;
int NbCollI = 0;
int NbColl = 0;

/* the user-defined function for the new operator */
void areequals(int *In, int *Inout, int const *Len, MPI_Datatype *Type) {
  int I;
  for (I = 0; I < *Len; I++) {
#ifndef NDEBUG
    // printf("Color %d called with color %d\n", *Inout, *In);
#endif
    if (*Inout != *In) {
      *Inout = ~0;
    }
    In++;
    Inout++;
  }
}

/* Check Collective MPI Function
 *
 *  color = type of collective (unique per collective)
 *  pcom = communicator
 *  OP_name = collective name
 *  OP_line = line in the source code of the collective
 *  warnings = warnings emitted at compile-time
 *  FILE_name = name of the file
 */
// NOLINTNEXTLINE
void check_collective_MPI(int OP_color, char const *OP_name, int OP_line,
                          char *Warnings, char *FileName) {
  int Rank;
  int SizeComm;

  // make sure MPI_Init has been called
  int Flag;
  MPI_Initialized(&Flag);

  if (Flag) {
    NbCc++;
    // Fortran programs are not handled
    MPI_Comm IniComm = MPI_COMM_WORLD;

    MPI_Comm_rank(IniComm, &Rank);
    MPI_Comm_size(IniComm, &SizeComm);

    int Res = 0;
    MPI_Op Equalsop;
    int Commutatif = 1;
    MPI_Op_create((void *)areequals, Commutatif, &Equalsop);

    MPI_Reduce(&OP_color, &Res, 1, MPI_INT, Equalsop, 0, IniComm);
    MPI_Op_free(&Equalsop);

    // Broadcast the result to everyone
    MPI_Bcast(&Res, 1, MPI_INT, 0, IniComm);

    if (Res == ~0) {
      if (strlen(Warnings) > 2) {
        printf("PARCOACH DYNAMIC-CHECK : Rank %d, warnings for my collective: "
               "%s\n",
               Rank, Warnings);
      }
      if (Rank == 0) {
        printf("PARCOACH DYNAMIC-CHECK : Error detected on rank %d\n"
               "PARCOACH DYNAMIC-CHECK : Abort is invoking line %d before "
               "calling %s in %s\n",
               Rank, OP_line, OP_name, FileName);
        MPI_Abort(MPI_COMM_WORLD, 0);
      }
    } else if (Rank == 0) {
      printf("PARCOACH DYNAMIC-CHECK : OK\n");
    }
  }
}

// NOLINTNEXTLINE
void check_collective_return(int OP_color, char const *OP_name, int OP_line,
                             char *Warnings, char *FileName) {
  int Rank;
  int SizeComm;

  // make sure MPI_Init has been called
  int flagend;
  int flagstart;
  MPI_Initialized(&flagstart);
  MPI_Finalized(&flagend);

  if (!flagend && flagstart) {
    // Fortran programs are not handled
    MPI_Comm IniComm = MPI_COMM_WORLD;

    MPI_Comm_rank(IniComm, &Rank);
    MPI_Comm_size(IniComm, &SizeComm);

    int Res = 0;
    MPI_Op Equalsop;
    int Commutatif = 1;
    MPI_Op_create((void *)areequals, Commutatif, &Equalsop);

    MPI_Reduce(&OP_color, &Res, 1, MPI_INT, Equalsop, 0, IniComm);
    MPI_Op_free(&Equalsop);

    // Broadcast the result to everyone
    MPI_Bcast(&Res, 1, MPI_INT, 0, IniComm);

    if (Res == ~0) {
      if (strlen(Warnings) > 2) {
        printf("PARCOACH DYNAMIC-CHECK : Rank %d, warnings for my collective: "
               "%s\n",
               Rank, Warnings);
      }
      if (Rank == 0) {
        printf("PARCOACH DYNAMIC-CHECK : Error detected on rank %d\n"
               "PARCOACH DYNAMIC-CHECK : Abort is invoking line %d before "
               "calling %s in %s\n",
               Rank, OP_line, OP_name, FileName);
        MPI_Abort(MPI_COMM_WORLD, 0);
      }
    } else if (Rank == 0) {
      printf("PARCOACH DYNAMIC-CHECK : OK\n");
    }
  }
}
