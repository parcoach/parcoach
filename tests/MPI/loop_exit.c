#include <stdlib.h>
#include <stdio.h>
#include "mpi.h"


int main(int argc, char** argv){
  int rank, size, i=1, j=10, var=0;

  MPI_Init(&argc,&argv);
  MPI_Comm_rank(MPI_COMM_WORLD,&rank);
  MPI_Comm_size(MPI_COMM_WORLD,&size);

  if(size<2){
    printf("This test needs at least 2 MPI processes\n");
    MPI_Finalize();
    return 1;
  }

  if(rank%2){
    while(i<10){
      i++;
    }
  }else{
    while(j<20){
      if (rank<10)
        exit(-1);
      j++;
    }
  }

  MPI_Barrier(MPI_COMM_WORLD);


  MPI_Finalize();
  return 0;
}
