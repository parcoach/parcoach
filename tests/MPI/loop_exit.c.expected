PARCOACH: tests/MPI/loop_exit.c: warning: MPI_Barrier line 31 possibly not called by all processes because of conditional(s) line(s)  19 (tests/MPI/loop_exit.c) 24 (tests/MPI/loop_exit.c) 25 (tests/MPI/loop_exit.c) (full-inter)
PARCOACH: tests/MPI/loop_exit.c: warning: MPI_Finalize line 33 possibly not called by all processes because of conditional(s) line(s)  19 (tests/MPI/loop_exit.c) 24 (tests/MPI/loop_exit.c) 25 (tests/MPI/loop_exit.c) (full-inter)
